core/src/test/java/com/linecorp/armeria/client/MaxNumEventLoopsPerEndpointTest.java

```diff
@@ -0,0 +1,307 @@
+        assertThat(entries4).hasSize(3); // Matched to Endpoint.of("a.com", 36462)
+
+        // Clear text SessionProtocols.
+
+        final List<AbstractEventLoopEntry> bComClearText =
+                s.entries(Endpoint.of("b.com", 80), SessionProtocol.H1C);
+        assertThat(bComClearText).hasSize(0);
+        acquireTenEntries(s, Endpoint.of("b.com"), SessionProtocol.H1C);
+        assertThat(bComClearText).hasSize(4); // Fallback to "b.com:80"
+
+        final List<AbstractEventLoopEntry> entries5 = s.entries(Endpoint.of("b.com"), SessionProtocol.H1C);
+        assertThat(bComClearText).isSameAs(entries5);
+
+        final List<AbstractEventLoopEntry> entries6 = s.entries(Endpoint.of("b.com"), SessionProtocol.H2C);
+        acquireTenEntries(s, Endpoint.of("b.com"), SessionProtocol.H2C);
+        assertThat(bComClearText).hasSize(4);
+        final List<AbstractEventLoopEntry> entries7 = s.entries(Endpoint.of("b.com"), SessionProtocol.HTTP);
+        assertThat(entries6).isSameAs(entries7);
+
+        // TLS SessionProtocols.
+
+        final List<AbstractEventLoopEntry> bComTls = s.entries(Endpoint.of("b.com", 443), SessionProtocol.H1);
+        assertThat(bComTls).hasSize(0);
+        acquireTenEntries(s, Endpoint.of("b.com"), SessionProtocol.H1);
+        assertThat(bComTls).hasSize(5); // Fallback to "b.com:433"
+
+        final List<AbstractEventLoopEntry> entries8 = s.entries(Endpoint.of("b.com"), SessionProtocol.H1);
+        assertThat(bComTls).isSameAs(entries8);
+
+        final List<AbstractEventLoopEntry> entries9 = s.entries(Endpoint.of("b.com"), SessionProtocol.H2);
+        acquireTenEntries(s, Endpoint.of("b.com"), SessionProtocol.H2);
+        assertThat(entries9).hasSize(5);
+        final List<AbstractEventLoopEntry> entries10 = s.entries(Endpoint.of("b.com"), SessionProtocol.HTTPS);
+        assertThat(entries9).isSameAs(entries10);
+
+        final List<AbstractEventLoopEntry> entries11 =
+                s.entries(Endpoint.of("b.com", 8443), SessionProtocol.H1);
+        assertThat(entries11).hasSize(1); // One entry is pushed when eventLoops.size() == maxNumEventLoops
+        acquireTenEntries(s, Endpoint.of("b.com", 8443), SessionProtocol.H1);
+        assertThat(entries11).hasSize(7); // No match
+    }
+
+    private static void acquireTenEntries(DefaultEventLoopScheduler s, Endpoint endpoint,
+                                          SessionProtocol sessionProtocol) {
+        for (int i = 0; i < 10; i++) {
+            s.acquire(endpoint, sessionProtocol);
+        }
+    }
+
+    /**
+     * eventLoops idx:   [   0,   1,   2,   3,   4,   5,   6]
+     *    - endpointA:       A    A
+     *    - endpointB:                 B    B    B
+     *    - endpointC:       C    C    C              C    C
+     *
+     * <P>The event loop group for endpoints is assigned sequentially. The endpointC needs 5 event loops so it
+     * takes the 3 event loops from the first 3 elements in the list.
```

Comment:
The endpointC starts at 5 requiring 5 event loops, so it wraps around at the end of the array and takes the first 3.